src/server/session.ts

```diff
@@ -239,9 +385,103 @@ namespace ts.server {
             }
         }
 
+        static matchFileByHash(file: string, index: DtsHistory, host: System): { fileName: string; path: string; commitsBehind: number; } {
+            // See if a file with this name is in the index
+            var lastSlash = Math.max(file.lastIndexOf('/'), file.lastIndexOf('\\'), -1);
+            var filenameOnly = file.substr(lastSlash + 1);
+            for (var i = 0; i < index.length; i++) {
+                if (index[i].n === filenameOnly) {
+                    // Just check for a filename match, ignore commit history
+                    return { fileName: filenameOnly, path: index[i].p, commitsBehind: -1 };
+                    
+                    // Alternatively, we could check against the .d.ts history and report how many commits behind it is
+                    /*var crc = dtsHash(host.readFile(file, 'utf-8'));
+                    for (var j = 0; j < index[i].h.length; j++) {
+                        if (crc === index[i].h[j]) {
+                            return { fileName: filenameOnly, path: index[i].p, commitsBehind: j };
+                        }
+                    }*/
+                }
+            }
+            return undefined;
+        }
+
+        private fetchingDts = false;
+        upToDateCheck(file: string, project: Project) {
+            if (this.projectService.getFormatCodeOptions(ts.normalizePath(file)).CheckForDtsUpdates === false) return;
+
+            var doUpToDateCheck = () => {
+                try {
+                    var info = Session.matchFileByHash(file, this.dtsVersionHistory, this.host);

```

Comment:
If we're not doing out-of-date checking, we shouldn't bother doing a hash-sensitive check of the file contents -- filename alone should be sufficient (and significantly faster).